### ISSUE TITLE
Custom implementation of ITelemetryInitializer doesn't receive ITelemetry of type RequestTelemetry in Isolated Azure Func(v4)

### ISSUE THREAD

**Issue by brewstertimothybrewster** [Azure/azure-functions-dotnet-worker#1603]
I have a Azure isolated function v4 and want to enrich the telemetry before it's sent to Application Insight. 
Therefore i have implemented the "ITelemetryInitializer" with a custom class. 
However when I execute a http request to the function the "Initialize" method is called with a "ITelemetry" object. 
But the problem is that none of these "ITelemetry" items are of type "RequestTelemetry". Only "ITelemetry" items of type "Microsoft.ApplicationInsights.DataContracts.MetricTelemetry" and "Microsoft.ApplicationInsights.DataContracts.DependencyTelemetry" are found.

![image](https://github.com/Azure/azure-functions-dotnet-worker/assets/73719180/35d35649-af4e-444a-b67c-a2f5a86d62ef)

My project is using these packages
![image](https://github.com/Azure/azure-functions-dotnet-worker/assets/73719180/889832e0-17ab-4c82-982d-7679abbfafee)

I would like to know if enrichment of telemetry is possible with Azure isolated function version 4. 
I have this working with another Asp.Net project btw.

**Comment by rtgs-danila-golovniov** [comment]
+1
we also want to enrich request telemetry with some correlation data, but are unable to do that from the worker process. We ended up with a custom middleware that logs a duplicate RequestTelemetry, which obviously at scale will cost **a lot**. So fixing this would be such a winner.

**Comment by angel199321** [comment]
> +1 we also want to enrich request telemetry with some correlation data, but are unable to do that from the worker process. We ended up with a custom middleware that logs a duplicate RequestTelemetry, which obviously at scale will cost **a lot**. So fixing this would be such a winner.

Hi,
Can you share the middleware codes, I trying to add custom correlation data to request telemetry too. 
Thank you.

**Comment by bmacalindong** [comment]
Just wanted to know what's the plan for this or are there are any alternatives we can use apart from tracking it via middleware?

**Comment by briandunnington** [comment]
would also like to hear an update about this. even if we cant get the real RequestTelemetry in the worker, it would be nice if the host "knew" about HTTP triggers and logged them as success/failures like people would expect. previously we had a custom ITelemetryInitializer just to do this:

![image](https://github.com/Azure/azure-functions-dotnet-worker/assets/1545776/799a60ca-6264-4a53-a738-fe277dc8b857)

at least with that, 4xx and 5xx responses show up as failed requests in AppInsights. otherwise, everything just gets logged as a success

**Comment by saalocin** [comment]
+1 experiencing exactly the same issue as described by OP.
The types received by `ITelemetryInitializer` is limited by these 2 types:
"Microsoft.ApplicationInsights.DataContracts.MetricTelemetry" and "Microsoft.ApplicationInsights.DataContracts.DependencyTelemetry"

**Comment by briandunnington** [comment]
Since the RequestTelemetry is handled entirely by the host process, and all of your function app code runs in the worker process, there is no straight-forward way to get access to the RequestTelemetry. You can in fact do it though, but it is not trivial. Full details of how to do it and a working example here: https://github.com/briandunnington/AzureFunctionsIsolatedHttpTelemetry (includes link to a published Nuget package if you only want to set the Success based on HTTP response codes)

**Comment by iceberg-sbourbeau** [comment]
One thing we'd like to be able to do is add properties to the RequestTelemetry
We're currently doing that with our InProc and it is of incredible help to debug.
For example, we have custom properties such as "CreditApplicationId" and "LoanId". Each function that manipulates those entities or subentity sets those identifier. Then, when an issue arises in production with one CreditApp, we can just filter the RequestTelemetry on the attribute "CreditApplicationId" (for example) and set the ID to get the complete flow (sometimes during multiple days) of what was done.

**Comment by basilfx** [comment]
@iceberg-sbourbeau If you custom properties using `Activity?.Curent.AddTags()`, they will appear on the 'Invoke' dependency telemetry emitted by the worker, just as they would on the request telemetry.

In general, we decided to ditch as much of the host logging as possible. We then added a custom telemetry processor that effectively takes the dependency telemetry emitted by the worker, and 'converts' it into a request telemetry. In order to do so, we had to add some middleware to catch the request URL, and use a bit of reflection to copy the telemetry context. Because the request telemetry is emitted by the worker, it can be further processed using other telemetry filters.

It's not perfect, but it solves two problems for us:

* Every one here is used to search for requests when starting an investigation.
* Having a request telemetry emitted by the host and a dependency telemetry feels duplicated. We now only emit one.

**Comment by luthus** [comment]
@fabiocav will this be picked up for the .NET 10 release? I'm seeing warnings on the Azure Portal that the in-process model will be retired on  November 10th 2026.

**Comment by RohitRanjanMS** [comment]
You're right—in-proc is being retired. We're looking into the feature that will allow worker process to add tags to the request telemetry emitted by the host process. I've heard plenty of feedback, and this seems like a good way to make things easier.

**Comment by RohitRanjanMS** [comment]
I have added a work item to address this - https://github.com/Azure/azure-functions-host/issues/11397

I will close this issue now, if you have any feedback, please add it in the above work item.
